### PR TITLE
8296975: RISC-V: Enable UseRVA20U64 profile by default

### PR DESCRIPTION
**Pull request by zhengxiaolinX**
The main purpose is to turn the option `UseRVC` on by default before JDK20 RDP 1. As per discussions [1], we can enable `UseRVA20U64`[2] by default to fulfill this.

```
> build/linux-riscv64-server-fastdebug/images/jdk/bin/java -XX:+PrintFlagsFinal -version | grep -E "UseRVC|UseRVA20U64"
     bool UseRVA20U64                              = true                                 {ARCH product} {default}
     bool UseRVC                                   = true                                 {ARCH product} {default}
openjdk version "20-internal" 2023-03-21
OpenJDK Runtime Environment (fastdebug build 20-internal-adhoc..jdk)
OpenJDK 64-Bit Server VM (fastdebug build 20-internal-adhoc..jdk, mixed mode)
```

[1] https://mail.openjdk.org/pipermail/riscv-port-dev/2022-November/000668.html
[2] https://github.com/openjdk/jdk/blob/873eccde01895de06e2216f6838d52d07188addd/src/hotspot/cpu/riscv/vm_version_riscv.cpp#L39-L44

Thanks,
Xiaolin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296975](https://bugs.openjdk.org/browse/JDK-8296975): RISC-V: Enable UseRVA20U64 profile by default

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) ⚠️ Review applies to [55e3dbe7](https://git.openjdk.org/jdk/pull/11155/files/55e3dbe73dc1a7ae1e072f99d98cf2eb60c6796e)
 * [Vladimir Kempik](https://openjdk.org/census#vkempik) (@VladimirKempik - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11155/head:pull/11155` \
`$ git checkout pull/11155`

Update a local copy of the PR: \
`$ git checkout pull/11155` \
`$ git pull https://git.openjdk.org/jdk pull/11155/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11155`

View PR using the GUI difftool: \
`$ git pr show -t 11155`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11155.diff">https://git.openjdk.org/jdk/pull/11155.diff</a>

</details>

### CODE REVIEW

**Review by RealFYang — COMMENTED** [review]
You might want to disable UseRVA20U64 when the C extension is not available at the same time.

**Comment by zhengxiaolinX** [comment]
> You might want to disable UseRVA20U64 when the C extension is not available at the same time.

Oh yes. I forgot that one. Will add the logic soon.

**Comment by zhengxiaolinX** [comment]
Tested by adding a `_features &= (~CPU_C);` after the aux vector check.

The result is
```
> build/linux-riscv64-server-fastdebug/images/jdk/bin/java -XX:+PrintFlagsFinal -version | grep -E "UseRVC|UseRVA20U64"
OpenJDK 64-Bit Server VM warning: RVC is not supported on this CPU
OpenJDK 64-Bit Server VM warning: RVA20U64 profile is not supported on this CPU
     bool UseRVA20U64                              = false                                {ARCH product} {default}
     bool UseRVC                                   = false                                {ARCH product} {default}
openjdk version "20-internal" 2023-03-21
OpenJDK Runtime Environment (fastdebug build 20-internal-adhoc..jdk)
OpenJDK 64-Bit Server VM (fastdebug build 20-internal-adhoc..jdk, mixed mode)
```

**Comment by luhenry** [comment]
Given the RVC is a _mandatory_ extension of `RVA20U64`, but also that `RVA20U64` can imply other extensions ([full list of mandatory extensions](https://github.com/riscv/riscv-profiles/blob/main/profiles.adoc#512-rva20u64-mandatory-extensions)), I wouldn't just disable `RVA20U64` when RVC isn't available, but I would fail completely the JVM. Getting in such case would mean that a hardware is implementing only parts of the __mandatory__ list of extensions for a given profile.

I understand that it puts the approach of this PR in question though, and I'm wondering why we need to enable a whole profile when we are trying to enable only one of the feature (RVC) and we also have feature flag detection for this feature (`_features & CPU_C`). It would just be better IMO to do proper feature detection and use that to enable/disable features, just like other platforms.

**Comment by zhengxiaolinX** [comment]
> 

Do you mean by directly using a solution like [1] without touching `UseRVA20U64`?

I think I am okay with any case.

[1] https://github.com/openjdk/jdk/pull/11155/commits/b5b9c64529c27c40542f8cda720652fabf70682d

**Comment by luhenry** [comment]
I mean something like https://github.com/openjdk/jdk/commit/d75b565dabc1dab3c508d2b4b83d34af5a1c7a35 (it hasn't been built or tested).

**Comment by zhengxiaolinX** [comment]
> I mean something like [d75b565](https://github.com/openjdk/jdk/commit/d75b565dabc1dab3c508d2b4b83d34af5a1c7a35) (it hasn't been built or tested).

Thanks for the explanation :-)  I see, and your solution looks far better than mine. But making `UseRVA20U64` as default true is an opinion from @VladimirKempik discussed in the mailing list, so I would also like to ask if he is okay.

(I guess the `vm_exit_during_initialization` would make `UseRVA20U64` hard to become a default option though, for Java should start normally anywhere.)

**Comment by VladimirKempik** [comment]
Well, did you see any rv64g hardware without Compressed Opcodes support ? 
Of course some may configure custom cpu core without RVC support, but it's usually done for custom environments where even linux is not present.
as was mentioned in ML, the rva20u64 profile also requires unaligned memory access support, which is still implicit requirement of jdk on risc-v ( see https://bugs.openjdk.org/browse/JDK-8291550 ), some cpus support it in hardware, some support m-mode emulator only.

Basically we are safe to enable rva20u64 by default I think. Nothing will change or break, and we can catch some RVC related bugs. if anything we can change that logic

**Review by VladimirKempik — APPROVED** [review]

**Comment by zhengxiaolinX** [comment]
Receiving no further comments, I think I am okay as well to remain the current PR.

@luhenry 's suggestion also provides good thoughts for me, to dispatch the handling logic to `UseRVA20U64` and `UseRVA22U64` themselves, when hardware does not support RVC. Thank you. And it appears to me that `UseRVA20U64` is your code, I guess you would like to see that it is a default true value.

I think we can move forward with this, and I need certainly to keep an eye on if there are issues after this PR.

Thank you all for the reviews and thoughts!

/integrate
